src/vivarium_census_prl_synth_pop/data/generate_business_names.py

```diff
@@ -0,0 +1,119 @@
+    real_but_uncommon_names = set(s_name_freq[s_name_freq < 1_000].index)
+
+    # Generate random business names.  Drop duplicates and overlapping names with uncommon names
+    n_total_names = int(n_total_names)  # Make int because of sys args
+    new_names = pd.Series()
+
+    # Generate additional names until desired number of random business names is met
+    while len(new_names) < n_total_names:
+        n_needed = n_total_names - len(new_names)
+        more_names = sample_names(
+            bigrams, n_needed, data_values.BUSINESS_NAMES_MAX_TOKENS_LENGTH
+        )
+        new_names = pd.concat([new_names, more_names]).drop_duplicates()
+        new_names = new_names.loc[~new_names.isin(real_but_uncommon_names)]
+
+    new_names.to_csv(
+        paths.BUSINESS_NAMES_DATA_ARTIFACT_INPUT_PATH,
+        header=["business_names"],
+        index=False,
+        compression="bz2"
+    )
+
+
+def make_bigrams(df: pd.DataFrame):
+    # Makes default dict of business names for map to sample from
+    # bigrams will be a Dict[str: Dict[str: int]]
+    # Example {"<start>": {keys are all first words for businesses: values are frequence where these pairs happen}}
+
+    def dict_factory():
+        return lambda: defaultdict(int)
+
+    bigrams = defaultdict(dict_factory())
+    n_rows = len(df)  # expect a few minutes
+
+    for i in range(n_rows):
+        if i % 10_000 == 0:
+            print(".", end=" ", flush=True)
+        names_i = df.iloc[i, 0]
+
+        tokens_i = names_i.split(" ")
+        for j in range(len(tokens_i)):
+            if j == 0:
+                bigrams["<start>"][tokens_i[j]] += 1
+            else:
+                bigrams[tokens_i[j - 1]][tokens_i[j]] += 1
+        bigrams[tokens_i[j]]["<end>"] += 1
+
+    return bigrams
+
+
+def sample_names(bigrams: defaultdict, n_businesses: int, n_max_tokens: int) -> pd.Series:
+    """
+
+    Parameters
+    ----------
+    bigrams: Default dict produced from make_bigrams function (see formatting of default dict.
+    n_businesses: Int of how many business names to generate
+    n_max_tokens: Int of max number of words possible for a business name to contain
+
+    Returns
+    -------
+    A string that is a randomly generated business name
+    """
+
+    logger.info("")
+    columns = [f"word_{i}" for i in range(n_max_tokens)]
+    names = pd.DataFrame(columns=columns)
+    names["word_0"] = ["<start>"] * n_businesses
+
+    for i in range(1, n_max_tokens):
+        if i == 1:
+            logger.info("Initializing random name sampling for business name generator")
+        previous_word = f"word_{i - 1}"
+        next_word = f"word_{i}"
+        current_words_count_dict = names[previous_word].value_counts().to_dict()
+        for word in current_words_count_dict.keys():
+            logger.info(
+                f"Generating {len(list(current_words_count_dict.keys()))} words "
+                f"in {i}th of {n_max_tokens} columns for business names data."
+            )
```

Comment:
This logging message could be improved.  It currently prints the cycles through the column and how many words are being created by the previous tokens.  I think adding it here or below after tokens is sampled in line 102 both make the messages printed out quite messy.

Reply:
yeah this might be more logging than we need. what is the purpose of the logging? are you wanting to track this as it runs?

Reply:
Yes.  Basically tracking progress as it generates names just since it will be running for a long time.